### ISSUE TITLE
Vimscript: new syntax \=l:var and \=a:var to expand the variable before calling the user-defined command-backend function

### ISSUE THREAD

**Issue by psprint** [neovim/neovim#12791]
## TL;DR; what's the feature-request is about:

1. A new syntax **`\=variable`** that'll would work exactly as **`deepcopy(variable)`**, i.e.: `call fun(\=a:arg)` would work as `call fun(deepcopy(a:arg))`.

2. The additional, however from the POV of the expected useful use-case **the most important property**: the **`\=`** operator should be active **EVEN** when used as an **`<f-args>`**, **`<q-args>`**, etc. -quoted-argument of a user-defined command, i.e.: e.g.: as in **`CMD \=l:var`** of a **`command -nargs=* CMD cmd_impl(<f-args>)`**-defined command.

3. It's possible that it would be best if the **`\=`** operator would **FULLY** follow its origin (i.e.: the place from which it is copied/imitated from) **←→** i.e.: [the **`\=`**-operator](https://vimhelp.org/change.txt.html#%3As%5C%3D) from the _s:/.../**&lt;into-substitute&gt;**/_ -argument of the **`:substitute`** command (and the `substitute()` function) **AND THUS BE**: followed by a **general _expression_**, not only by a **_variable_**.

The remaining text describes the usefulness of the main-enabled use-case: the ability to easily pass **local variables** into the **backend-functions** of **user-defined** commands **EVEN** when passing them quoted as **`<f-args>`**, **`<q-args>`**, etc. → by simply preceding them with **`\=...`** when calling the command, as in the example of the above item **`(2)`**.

### Is your feature request about something that is currently impossible or hard to do? Please describe the problem.

I have a very useful (debug-)print command that **expands any detected _g:vars_, _s:vars_ and _b:vars_** in the obtained as-is arguments-text:

```vim
command! -nargs=+ -count=4 -bang -bar PRINT call s:print_impl(<count>,<q-bang>,expand("<sflnum>"),<f-args>)
```

As it can be seen:
- the PRINT command takes variable # of args,
- it recognizes and accepts the bang-symbol,
- it recognizes and accepts the count-prefix, which is set to 4 by default (the message-level / the default highlight-color of it),
- it can be followed by a follow-up bar-command: **PRINT ... | cmd2 ...**,
- it passes the script-line number of the place that it's called via `expand("<sflnum>")`.

More, this time the more feature-request related details:
- the PRINT command quotes the variadic args by using the: **`<f-args>`**-symbol,
- thus, when a variable-name, like `g:my_var` in `PRINT g:my_var` will be given, **it'll be passed to the implementation-function `print_impl()` as a raw, unexpanded text: `g:my_var`**.

This isn't a big problem, as such strings (like `g:var` or `s:var`, etc.) can be easily matched in `print_impl()` and expanded by hand (via `eval()` or via `get(g:,matched_variable_name, ...)`, etc.). There's however one drawback of this approach -- **THE IMPOSSIBLE-TO-SOLVE ROOT-PROBLEM OF THIS FEATURE REQUEST** -- the inability to expand the calling-function local variables `l:var` and `a:var` within the under-scope of `print_impl()` function. Thus, there is no way to make this work:

```vim
   let my_var = "Hi!"
   " PROBLEM: Cannot expand l:my_var in the command-backend function -- the variable isn't defined inside its scope...
   PRINT l:my_var
```

as the local-variable `l:my_var` isn't declared/accessible inside the `print_impl()` function which receives the text `"l:my_var"`.

The same happens when trying to print an argument-type variable `a:var`:

```vim
function fun(my_function_argument)
   " ...
   " PROBLEM: Cannot expand a:my_function_argument in the command-backend function -- it's undeclared there...
   PRINT a:my_function_argument
   " ...
```

### Describe the solution you'd like

**LITTLE BACKGROUND:** I think that such a convenient printing-function that **doesn't need no quoting at all** of it's input, e.g.:

```vim
   PRINT Hi toupper("there")! -- yes, I also detect function-call syntax and expand it via \eval() unless quoted with b-slash
   PRINT The variable is s:variable, it exists: exists("s:variable")
   7PRINT Different message-level/color #7, also (toupper("i") . " concatenate" . " arguments wrapped in parens")
   PRINT! So (2 + 3) is 5, while (toupper("hi") . " there") is "HI there" – as an error message (because of the bang!)
```

is a very useful and convenient thing. It would be great if using it **WOULD NOT enforce promoting local-variables to session-ones**, i.e.:

```vim
   let l:var = "This cannot work, so ...^H^H^H..."
   let s:var = "... I declare var as a session-type variable to make it accessible inside print_impl() for expanding..."
   PRINT The variable is s:var, an enforced session-type kind :( ... Polluting the namespace :( ...
```

That's the main pain behind this request.

**THE PROPOSED SOLUTION IS** a variable-forwarding syntax `\=l:variable` and `\=a:variable` that **makes the variable given after the `\=` accessible within the scope of the called function**, i.e.:

```vim
" An example call-situation:
function Func(forwarded_arg)
   let var = 'SUPER! I will be visible/accessible inside print_impl()!`
   let a:forwarded_arg = 'EXTRA! I will be also forwarded and visible/accessible in the underlying scope of print_impl()!'
   PRINT The var is: \=l:var, while the argument is \=a:forwarded_arg !
endfunction

" An example backend-function of the PRINT command:
function Print_Impl(...)
    let input = deepcopy(a:000)
    let idx = 0
    for text in input
        " l:var and a:forwarded_arg are simply accessible here because they're forwarded
        if text =~ '\v^[sgbla]:[a-zA-Z0-9_]+$' | let input[idx]=eval(text) | endif
        let idx += 1
    endfor
    " Output the final message with all the variables (given *by-name*) expanded
    echom join(input)
endfunction

" Define the command. It'll have all its input quoted with " and passed down as-is, by the use of the <f-args> operator:
command! -nargs=+ -bar PRINT call Print_Impl(<f-args>)
```

### Describe alternatives you've considered

I've also considered that the **`\=variable`** syntax instead of **forwarding** of **variable** into the called-function scope should  **force a value-copying of it before passing it to the function (the _print_impl()_ in the examples)**. So this would would work just as if preceding a variable with the new operator i.e.: e.g.: **`\=l:var`** would be doing **`deepcopy(l:var)`**. Its use-cases would be IMHO: **(a)** to induce the value-fetching before quoting with **`<f-args>`** in the user-defined commands, **(b)** a **compact and nice shorthand for `deepcopy(variable)`**.

I'm not completely sure which of the two variants of **`\=`** (the scope-forwarding **or** value-copying) should be implemented. The description fits/explains both variants well. I would now, after writing the (1st-variant oriented) description, opt for the second-type variant, i.e.: the **`\=...` ←→ `deepcopy(...)`** as it's a **(a)** more basic, simple function, **(b)** a quite useful general, not only user-defined commands -related feature (i.e.: the convenient `deepcopy(...)` shorthand), **(c)** it's easy to implement AND also an IMO very consistent, VimL shape and condition retaining-extension of the language (because it's just a **deepcopy(...)** and because it's-consistent/it-imitates [the **`\=`**-operator](https://vimhelp.org/change.txt.html#%3As%5C%3D) from **:substitute**).

In short, the 2nd variant of the **`\=var_name`** operator is:

### 1. **Cause the `var_name` to be used as a to-copy _value_ of it** (just like if preceding with **`\=var`** would just cause **`deepcopy(var)`** to occur) and **«EVEN»** when used in on a place of any **`<f-args>`-passed argument of a `:command`-defined user command**. 

**And that the above item: _(1)_ is the actual feature-request of this issue also summed up in the TL;DR; section.** The description well serves also this (second) variant (all content except for the unneeded `Print_Impl()` function example) so I'll leave it unchanged (for now).

### Additional context

- I think that the choice of **`\=`** is a very lucky one, as it follows/mimics the **`\=`** operator for the result/to-substitute `substitute` argument, where it's meaning is: *what follows is a sub-expression, evaluate it*. Here, it'll be mimicked as: **what follows is a variable, evaluate and return its value**.
- The source code of the PRINT implementation-function can be found [here](https://github.com/zphere-zsh/vim-user-menu/blob/025a51bd1950abec801f736dadee2606d8eb1165/plugin/vim-user-menu.vim#L339).

**Comment by justinmk** [comment]
Thanks for your thoughtful request.

> ```viml
>    let my_var = "Hi!"
>    " PROBLEM: Cannot expand l:my_var in the command-backend 
>    " function -- the variable isn't defined inside its scope...
>    PRINT l:my_var
> ```

Would be another layer in Vimscript's execution model...

> **THE PROPOSED SOLUTION IS** a variable-forwarding syntax `\=l:variable` and `\=a:variable` that **makes the variable given after the `\=` accessible within the scope of the called function**, i.e.:
> 
> ```viml
> " An example call-situation:
> function Func(forwarded_arg)
>    let var = 'SUPER! I will be visible/accessible inside print_impl()!`
>    let a:forwarded_arg = 'EXTRA! I will be also forwarded and visible/accessible in the underlying scope of print_impl()!'
>    PRINT The var is: \=l:var, while the argument is \=a:forwarded_arg !
> endfunction
> ```


Seems like an [x-y problem](https://en.wikipedia.org/wiki/XY_problem). Your example shows a `PRINT` command where you want to interleave Vimscript variables, to avoid the need to call a function, presumably because typing function parameters is a chore:

    call printf('The var is: %s while the argument is %s', l:var, a:forwarded_arg)

But Vim has numerous features to fix the chore of typing. There is [printf plugin](https://github.com/mptre/vim-printf) for example.


- More generally, we are focused on enhancing Lua usage in Nvim. We aren't going to spend much time enhancing Vimscript.
- related https://github.com/vim/vim/issues/6791
